sycl/plugins/unified_runtime/ur/adapters/cuda/image.cpp

```diff
@@ -0,0 +1,1029 @@
+      } else if (pImageDesc->type == UR_MEM_TYPE_IMAGE3D) {
+        // Cannot create 3D image from USM.
+        return UR_RESULT_ERROR_INVALID_VALUE;
+      }
+    } else {
+      // Unknown image memory type.
+      return UR_RESULT_ERROR_INVALID_VALUE;
+    }
+
+    RetErr = urTextureCreate(hContext, hSampler, pImageDesc, image_res_desc,
+                             phImage);
+
+    auto urMemObj = std::unique_ptr<ur_mem_handle_t_>(new ur_mem_handle_t_{
+        hContext, (CUarray)hImageMem, (CUtexObject)*phImage, hSampler,
+        pImageDesc->type});
+
+    if (urMemObj == nullptr) {
+      return UR_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    *phMem = urMemObj.release();
+  } catch (ur_result_t Err) {
+    return Err;
+  } catch (...) {
+    return UR_RESULT_ERROR_UNKNOWN;
+  }
+
+  return RetErr;
+}
+
+UR_APIEXPORT ur_result_t UR_APICALL urBindlessImagesImageCopyExp(
+    ur_queue_handle_t hQueue, void *pDst, void *pSrc,
+    const ur_image_format_t *pImageFormat, const ur_image_desc_t *pImageDesc,
+    ur_exp_image_copy_flags_t imageCopyFlags, ur_rect_offset_t srcOffset,
+    ur_rect_offset_t dstOffset, ur_rect_region_t copyExtent,
+    ur_rect_region_t hostExtent, uint32_t numEventsInWaitList,
+    const ur_event_handle_t *phEventWaitList, ur_event_handle_t *phEvent) {
+  UR_ASSERT(hQueue, UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+  UR_ASSERT(pDst, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+  UR_ASSERT(pSrc, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+  UR_ASSERT(pImageFormat, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+  UR_ASSERT(pImageDesc, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+  UR_ASSERT((imageCopyFlags == UR_EXP_IMAGE_COPY_FLAG_HOST_TO_DEVICE ||
+             imageCopyFlags == UR_EXP_IMAGE_COPY_FLAG_DEVICE_TO_HOST ||
+             imageCopyFlags == UR_EXP_IMAGE_COPY_FLAG_DEVICE_TO_DEVICE),
+            UR_RESULT_ERROR_INVALID_VALUE);
+
+  ur_result_t RetErr = UR_RESULT_SUCCESS;
+
+  unsigned int NumChannels = 0;
+  size_t PixelTypeSizeBytes = 0;
+
+  RetErr = UR_CHECK_ERROR(
+      urCalculateNumChannels(pImageFormat->channelOrder, &NumChannels));
+
+  // We need to get this now in bytes for calculating the total image size
+  // later.
+  RetErr = UR_CHECK_ERROR(urToCudaImageChannelFormat(
+      pImageFormat->channelType, nullptr, &PixelTypeSizeBytes));
+
+  size_t PixelSizeBytes = PixelTypeSizeBytes * NumChannels;
+
+  try {
+    ScopedContext Active(hQueue->getContext());
+    CUstream Stream = hQueue->getNextTransferStream();
+    RetErr =
+        enqueueEventsWait(hQueue, Stream, numEventsInWaitList, phEventWaitList);
+    // We have to use a different copy function for each image dimensionality.
+
+    if (imageCopyFlags == UR_EXP_IMAGE_COPY_FLAG_HOST_TO_DEVICE) {
+      if (pImageDesc->type == UR_MEM_TYPE_IMAGE1D) {
+        size_t CopyExtentBytes = PixelSizeBytes * copyExtent.width;
+        char *SrcWithOffset = (char *)pSrc + (srcOffset.x * PixelSizeBytes);
+        RetErr = UR_CHECK_ERROR(
```

Comment:
The status of `RetErr` as set by `enqueueEventsWait` will be lost here, without it being checked, is that safe to do?

Reply:
I'm not sure.. Nonetheless I've removed it and rely on `UR_CHECK_ERROR` for it.